### ISSUE TITLE
Differing union order results between browsers with a custom function

### ISSUE THREAD

**Issue by ymor** [FontoXML/fontoxpath#252]
Hi,
We have this concept of cross document querying.  I've created the following as an example: https://codesandbox.io/s/mystifying-lalande-bxh5j?file=/src/App.tsx

We recently came across a scenario where we needed to do a union across these documents but the results are returning in a different order between chrome and firefox.  Now I'm not sure on the rules for the return values order - I noticed in the same document its document order. But I would have thought it should be consistent between browsers.

It's calling the functions in the same order, just the results get re-ordered.
Thanks.

**Comment by DrRataplan** [comment]
Cool, good find! Most likely, this is because chrome and firefox have a slightly different implementation of the sort function, one is stable, the other less so. 

https://www.w3.org/TR/xquery-31/#id-document-order:
> The relative order of nodes in distinct trees is stable but implementation-dependent, subject to the following constraint: If any node in a given tree T1 is before any node in a different tree T2, then all nodes in tree T1 are before all nodes in tree T2.

I think the behaviour you are seeing is not the best. We can fix it by implementing our own sort (I'd say a merge sort). The code lives in documentOrderUtils.ts. To be honest, I do not care personally whether it should be stable or not, as long as it behaves the same in both browsers. Would you feel confident in making a PR to do this?

**Comment by ymor** [comment]
Sure will take a look. Thanks for pointing me in the right direction!

**Comment by ymor** [comment]
Hi,
So I've managed to resolve this with a merge sort. https://github.com/FontoXML/fontoxpath/compare/master...ymor:%23252-custom-function-union-order
However I'm not sure on the best way to test the original issue. I did add some headless browser test support: https://github.com/FontoXML/fontoxpath/compare/master...ymor:%23252-custom-function-union-order-with-browser-test
But not sure if that's a route you want to go down?  Happy to add a normal mocha test if that's going to be good enough? Also happy to move the merge sort function/tests to a more appropriate place.
Thanks.

**Comment by DrRataplan** [comment]
That looks cool, the browser tests were something we had, but we removed it when we went from plain closure-compiled JS to closure-compiled TS, just to make the build system easier to maintain. Seeing that the tests do not require that much configuration, I think it would be cool to re-add them and run them in Travis on each commit, alongside all of the other tests in Node (QT3, XQUT, etc).

Note that your new test file is indented with spaces though. I think the location is fine as it is, we'll move it to another place once we find out we can use it in more places.

We have a new micro-performance tool integrated in FontoXPath, I'll run it on your new code somewhere today to verify the performance of the manually implemented sort is comparable to the built-in sort.

Thanks!
